File: easy/108. Convert Sorted Array to Binary Search Tree/108. Convert Sorted Array to Binary Search Tree.ts
/**
 * Definition for a binary tree node.
 */

class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

function sortedArrayToBST(nums: number[]): TreeNode | null {
    if (!nums.length)
        return null

    const middle = ~~(nums.length / 2)
    const left = sortedArrayToBST(nums.slice(0, middle))
    const right = sortedArrayToBST(nums.slice(middle + 1,))
    return new TreeNode(nums[middle], left, right)
};

File: easy/108. Convert Sorted Array to Binary Search Tree/108. Convert Sorted Array to Binary Search Tree.go
package main

import "math"

/**
 * Definition for a binary tree node.
 */

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func sortedArrayToBST(nums []int) *TreeNode {
	if len(nums) == 0 {
		return nil
	}

	middle := int(math.Round(float64(len(nums) / 2)))
	left := sortedArrayToBST(nums[middle:])
	right := sortedArrayToBST(nums[middle+1:])
	return &TreeNode{Val: nums[middle], Left: left, Right: right}
}
